src/lib/doctrine-dbal-bulk/src/Flow/Doctrine/Bulk/Dialect/MySQLDialect.php

```diff
@@ -0,0 +1,105 @@
+use Flow\Doctrine\Bulk\TableDefinition;
+
+final class MySQLDialect implements Dialect
+{
+    /**
+     * @param TableDefinition $table
+     * @param BulkData $bulkData
+     * @param array{
+     *  skip_conflicts?: boolean,
+     *  upsert?: boolean,
+     *  update_columns?: array<string>
+     * } $insertOptions $insertOptions
+     *
+     * @return string
+     */
+    public function prepareInsert(TableDefinition $table, BulkData $bulkData, array $insertOptions = []) : string
+    {
+        if (\array_key_exists('skip_conflicts', $insertOptions) && $insertOptions['skip_conflicts'] === true) {
+            return \sprintf(
+                'INSERT INTO %s (%s) VALUES %s ON DUPLICATE KEY UPDATE %4$s=%4$s',
+                $table->name(),
+                $bulkData->columns()->concat(','),
+                $bulkData->toSqlPlaceholders(),
+                \current($bulkData->columns()->all())
+            );
+        }
+
+        if (\array_key_exists('upsert', $insertOptions) && $insertOptions['upsert'] === true) {
+            return \sprintf(
+                'INSERT INTO %s (%s) 
+                VALUES %s 
+                ON DUPLICATE KEY UPDATE %s',
+                $table->name(),
+                $bulkData->columns()->concat(','),
+                $bulkData->toSqlPlaceholders(),
+                \array_key_exists('update_columns', $insertOptions) && \count($insertOptions['update_columns'])
+                    ? $this->updateSelectedColumns($insertOptions['update_columns'], $bulkData->columns())
+                    : $this->updateAllColumns($bulkData->columns())
+            );
+        }
+
+        return \sprintf(
+            'INSERT INTO %s (%s) VALUES %s',
+            $table->name(),
+            $bulkData->columns()->concat(','),
+            $bulkData->toSqlPlaceholders()
+        );
+    }
+
+    /**
+     * @param TableDefinition $table
+     * @param BulkData $bulkData
+     * @param array $updateOptions
+     *
+     * @throws RuntimeException
+     *
+     * @return string
+     */
+    public function prepareUpdate(TableDefinition $table, BulkData $bulkData, array $updateOptions = []) : string
+    {
+        return \sprintf(
+            'REPLACE INTO %s (%s) VALUES %s',
+            $table->name(),
+            $bulkData->columns()->concat(','),
+            $bulkData->toSqlPlaceholders()
+        );
+    }
+
+    /**
+     * @param Columns $columns
+     *
+     * @return string
+     */
+    private function updateAllColumns(Columns $columns) : string
+    {
+        return \implode(
+            ',',
+            $columns->map(
+                fn (string $column) : string => "{$column} = VALUES({$column})"
+            )
+        );
+    }
+
+    /**
+     * @param array<string> $updateColumns
+     * @param Columns $columns
+     *
+     * @return string
+     */
+    private function updateSelectedColumns(array $updateColumns, Columns $columns) : string
+    {
```

Comment:
I think I would use the same approach as above except that I would first filter Columns to only have updateColumns